utils/s2n_fork_detection.c

```diff
@@ -107,7 +107,7 @@ static inline S2N_RESULT s2n_initialise_wipeonfork_best_effort(void *addr, long
 static inline S2N_RESULT s2n_initialise_inherit_zero(void *addr, long page_size)
 {
 #if defined(S2N_MINHERIT_SUPPORTED) && defined(MAP_INHERIT_ZERO)
-    RESULT_ENSURE(minherit(addr, pagesize, MAP_INHERIT_ZERO) == 0, S2N_ERR_FORK_DETECTION_INIT);
```

Comment:
Well this is very concerning. Do you know why we never built this?

Reply:
I do not know why.
It's strange this condition exists but seems to have never been tested and doesn't get built on any of the more popular platforms. Same case with the delta on line 341.

Reply:
Thanks for this discovery!

First, let's make this clear: s2n-tls uses always-on prediction resistance. There is no escaped issue here even when, in some sub-cases, this piece of code accidentally didn't enter into the build.

Now, for the reason this code was removed during pre-compilation.

Apparently, FreeBSD doesn't define `MAP_INHERIT_ZERO`, but instead define `INHERIT_ZERO`. The implementation of the fork-zeroisation feature was done here: https://reviews.freebsd.org/D427. Hence, `MAP_INHERIT_ZERO` would never be defined on FreeBSD.

To verify this, I spun up a FreeBSD EC2 instance using FreeBSD 13.1. I made the following change to the s2n-tls source code:
```
diff --git a/utils/s2n_fork_detection.c b/utils/s2n_fork_detection.c
index 546a7a82..5d838657 100644
--- a/utils/s2n_fork_detection.c
+++ b/utils/s2n_fork_detection.c
@@ -31,6 +31,14 @@ typedef __darwin_pthread_once_t pthread_once_t;
 
 #include <sys/mman.h>
 
+#if defined(MAP_INHERIT_ZERO)
+#error "MAP_INHERIT_ZERO defined!"
+#endif
+
+#if defined(INHERIT_ZERO)
+#error "INHERIT_ZERO defined!"
+#endif
+
 /* Not always defined for Darwin */
 #if !defined(MAP_ANONYMOUS)
     #define MAP_ANONYMOUS MAP_ANON
```

I then ran the build which produced:
```
root@freebsd:/home/ec2-user/s2n-tls/s2n-build # ninja
[5/486] Building C object CMakeFiles/s2n.dir/utils/s2n_fork_detection.c.o
FAILED: CMakeFiles/s2n.dir/utils/s2n_fork_detection.c.o 
/usr/bin/cc -D_POSIX_C_SOURCE=200809L -I/home/ec2-user/s2n-tls -I/home/ec2-user/s2n-tls/api -pedantic -
[....]
CMakeFiles/s2n.dir/utils/s2n_fork_detection.c.o -c /home/ec2-user/s2n-tls/utils/s2n_fork_detection.c
/home/ec2-user/s2n-tls/utils/s2n_fork_detection.c:39:2: error: "INHERIT_ZERO defined!"
#error "INHERIT_ZERO defined!"
 ^
1 error generated.
```

A'ha! We need to cater for this FreeBSD-speciality in the code...

Finally, the reason this was not detected in the CI.

There are unit tests that verifies that each fork detection method works in isolation. However, the unit test first have to check whether the fork detection method is supported (otherwise, it doesn't make sense to test the feature, obviously...). But, the function asserting whether the minherit fork detection method is supported uses the existence of `MAP_INHERIT_ZERO` as a [condition](https://github.com/aws/s2n-tls/blob/main/utils/s2n_fork_detection.c#L338-L345). This means that tests that verifies this fork detection method in isolation are [skipped](https://github.com/aws/s2n-tls/blob/main/tests/unit/s2n_fork_generation_number_test.c#L305). 

Note that the CI also runs ALL unit tests on the default configuration. "Configuration" is really a bad choice of words, because you can't change this without source code changes. So, there is no escaped issue.

Reply:
For solution: https://github.com/aws/s2n-tls/issues/3674

Reply:
I think code coverage should help us diagnose these kinds of issues. I wonder if there is code build coverage -- to tell you which lines actually get built.